File: projects/fractals/php_mandelbrot/chaos.php
<?php
/*
function fac($n) {
	if($n==0){return 1;}
	elseif($n==1){return 1;}
	elseif($n==2){return 2;}
	elseif($n==3){return 6;}
	elseif($n==4){return 24;}
	elseif($n==5){return 120;}
	elseif($n==6){return 720;}
	else
	{
		return $n*fac($n-1);
	}
}


function npr($n, $p) {
if($n<$p){$xx=$n;$n=$p;$p=$xx;}
if($p==0){return 1;}
if($n==$p){return 1;}
return fac($n)/(fac($p)*fac($n-$p));
}

function is_outside_n($rec, $imc, $n, $level) {
	$im=$imc;
	$re=$rec;

	for($j=0;$j<$level;++$j) {
		if($re*$re+$im*$im > 256){
			return $j;
		}
		$reold=$re;
		$imold=$im;
		for($k=0;$k<=$n;++$k) {
			if($k%4==0) {$re+=npr($n, $k)*pow($reold, $n-$k)*pow($imold, $k);}
			elseif($k%4==1) {$im+=npr($n, $k)*pow($reold, $n-$k)*pow($imold, $k);}
			elseif($k%4==2) {$re-=npr($n, $k)*pow($reold, $n-$k)*pow($imold, $k);}
			elseif($k%4==3) {$im-=npr($n, $k)*pow($reold, $n-$k)*pow($imold, $k);}
		}
	}
	return $j;
}
*/
function is_outside($rec, $imc, $level) {

	$im=$imc;
	$re=$rec;

	for($j=0;$j<$level;++$j){
		$re2=$re*$re;
		$im2=$im*$im;
		if($re2+$im2 > 256){
			return $j;
		}
		$im=2*$re*$im+$imc;
		$re=$re2-$im2+$rec;
	}
	return $j;
}

// width of output image, square
if(!isset($_GET['wx'])){$wx=1024/10;}else{$wx=$_GET['wx'];}
if(!isset($_GET['wy'])){$wy=768/10;}else{$wy=$_GET['wy'];}

//Bild initialisieren
$image=imagecreate($wx,$wy);
// fill with background color, black
imagefill($image,0,0,0);
if(!isset($_GET['r'])){$r=2.0;}else{$r=$_GET['r'];}
if(!isset($_GET['recen'])){$recen=0.0;}else{$recen=$_GET['recen'];}
if(!isset($_GET['imcen'])){$imcen=0.0;}else{$imcen=$_GET['imcen'];}
$s=2.0*$r/$wx;

// generate color palette from black to white
$pi = 3.1416;
$steps = 256/6;
for($i=0,$rr=0,$gg=0,$bb=0; $i<255; ++$i) {
//schönes grau
// 	$xx = 255*abs(sin(2*$pi*$i/255));
// 	$yy = 255*abs(sin(2*$pi*$i/255));
// 	$zz = 255*abs(sin(2*$pi*$i/255));
//alien-grün
// 	$rr = 255*abs(sin($pi*$i/255))*exp(-$i); $rr%=255;
// 	$gg = 255*abs(sin($pi*$i/255)); $gg%=255;
// 	$bb = 255*abs(sin($pi*$i/255))*exp(-$i); $bb%=255;
//Regenbogen (?)
// 	if($i<1*$steps){ $dirr = 0;$dirg = +1;$dirb = 0;}
// 	elseif($i<2*$steps){$dirr=-1;$dirg=0;$dirb=0;}
// 	elseif($i<3*$steps){$dirr=0;$dirg=0;$dirb=1;}
// 	elseif($i<4*$steps){$dirr=0;$dirg=-1;$dirb=0;}
// 	elseif($i<5*$steps){$dirr=1;$dirg=0;$dirb=0;}
// 	elseif($i<6*$steps){$dirr=0;$dirg=0;$dirb=-1;}
// 
// 	$rr += $dirr*$steps; $gg += $dirg*$steps; $bb += $dirb*$steps;

	$rr = 255*abs(sin(2*$pi*$i/255)*exp(-$i));
	$gg = 255*abs(sin(2*$pi*$i/255));
	$bb = 255*abs(sin(2*$pi*$i/255)*exp(-$i));
	$col[$i] = imagecolorallocate($image,$rr,$gg,$bb);
}
for($y=0;$y<$wy;$y++){
	for($x=0;$x<$wx;$x++){
		$rec=1.0*$s*($x-$wx/2)+$recen;
		$imc=1.0*$s*($y-$wy/2)+$imcen;
// 		$scale=255-is_outside($rec, $imc, 255*(-log10($s)/2));
		$scale=255-is_outside($rec, $imc, 255);
// 		imagesetpixel($image, $x, $y, $col[($scale)%255]);
		imagesetpixel($image, $x-$scale/256, $y+$scale/256, $col[($scale)%255]);
	}
}

//Bild ausgeben
header ( 'Content-Type: image/gif' );
imagegif ( $image );


?>

File: projects/fractals/php_mandelbrot/src/chaos.cpp
// (c) 2008 GPL v3 Pascal & Benedikt Steger

// This program gives a file with an array as output, where every number corresponds the iteration after which z->z^2+c leaves radius 2

#include <iostream>
#include <fstream>
#include <cmath>
#include <QImage>

#define PI 3.14159265358

int fac(int n) {
	if(n==0){return 1;}
	else if(n==1){return 1;}
	else if(n==2){return 2;}
	else if(n==3){return 6;}
	else if(n==4){return 24;}
	else if(n==5){return 120;}
	else if(n==6){return 720;}
	else
		return n*fac(n-1);
}

int npr(int n, int p) {
	if(n<p){int xx=n;n=p;p=xx;}
	if(p==0){return 1;}
	if(n==p){return 1;}
	return fac(n)/(fac(p)*fac(n-p));
}

int is_outside_n(double rec, double imc, int  n, int level) {
	double im=imc;
	double re=rec;
	int j;
	for(j=0;j<level;++j) {
		if(re*re+im*im > 256){
			return j;
		}
		int reold=re;
		int imold=im;
		for(int k=0;k<=n;++k) {
			if(k%4==0) {re+=npr(n, k)*pow(reold, n-k)*pow(imold, k);}
			else if(k%4==1) {im+=npr(n, k)*pow(reold, n-k)*pow(imold, k);}
			else if(k%4==2) {re-=npr(n, k)*pow(reold, n-k)*pow(imold, k);}
			else if(k%4==3) {im-=npr(n, k)*pow(reold, n-k)*pow(imold, k);}
		}
	}
	return j;
}

int is_outside(double rec, double imc, int level) {
	double im=imc;
	double re=rec;
	int j;
	for(j=0;j<level;++j){
		double re2=re*re;
		double im2=im*im;
		if(re2+im2 > 256){
			return j;
		}
		im=2*re*im+imc;
		re=re2-im2+rec;
	}
	return j;
}

QRgb colfun(int i) {
    int rr = 255*std::abs(std::sin((3.14159265358*i)/255));//*std::exp(-i/10.0);    rr%=255;
    int gg = 255*std::abs(std::sin((3.14159265358*i)/255));            gg%=255;
    int bb = 255*std::abs(std::sin((3.14159265358*i)/255));//*std::exp(-i/10.0);    bb%=255;

    return qRgb(rr,gg,bb);
}

int main(int argc, char* argv[]){
    if(argc!=6) {
	std::cout << "usage: ./chaos r wx re im" << std::endl;
	std::cout << " with r:  radius of calculated area " << std::endl;
	std::cout << "      wx: width of the resulting image in pixels" << std::endl;
	std::cout << "      wy: height of the resulting image in pixels" << std::endl;
	std::cout << "      re: real part and" << std::endl;
	std::cout << "      im: imaginary part of the central point." << std::endl;
	return 1;
    }

double r=atof(argv[1]);
double wx = atoi(argv[2]);
double wy = atoi(argv[3]);
double recen = atof(argv[4]);
double imcen = atof(argv[5]);

double s=2.0*r/wx;

QImage image(wx, wy, QImage::Format_RGB32);

//std::ofstream fout("/home/bs/public_html/chaoso.php");
//fout << "<?php" << std::endl;
//fout << " $output=array();" << std::endl;
for(int y=0;y<wy;y++){
    //std::cout << " Calculating line " << y << " of " << wy << std::endl;
    //fout << "$output[" << y << "]=array(";
	for(int x=0;x<wx;x++){
		double rec=1.0*s*(x-wx/2)+recen;
		double imc=1.0*s*(y-wy/2)+imcen;
// 		scale=255-is_outside(rec, imc, 255*(-log10(s)/2));
		int scale=255-is_outside(rec, imc, 255);
 		image.setPixel(x, y, colfun(scale));
//		fout << scale << ",";
	}
//	fout << ");"<< std::endl;
}

image.save("out.png");
//fout << "?>" << std::endl;
//fout.close();
//std::cout << " done!" << std::endl;
}
